drivers/i2c/i2c_ll_stm32_v2.c

```diff
@@ -933,24 +945,51 @@ static int stm32_i2c_msg_write(const struct device *dev, struct i2c_msg *msg,
 			if (check_errors(dev, __func__)) {
 				return -EIO;
 			}
+
+			if ((k_uptime_get_32() - start_time) >
+			    STM32_I2C_TRANSFER_TIMEOUT_MSEC) {
+				goto error;
+			}
 		}
 
 		LL_I2C_TransmitData8(i2c, *buf);
 		buf++;
 		len--;
 	}
 
-	return msg_done(dev, msg->flags);
+	ret = msg_done(dev, msg->flags);
+	if (ret < 0) {
+		goto error;
+	}
```

Comment:
It didn't disable the peripheral when the `msg_done` returns an error. Also, it is not done in `stm32_i2c_msg_read` for the `msg_done` error.

Reply:
@erwango I'm going to do it later today